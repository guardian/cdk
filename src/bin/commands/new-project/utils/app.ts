import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

interface AppBuilderProps {
  imports: Imports;
  appName: Name;
  stack: Name;
  stages: string[];
  outputFile: string;
  outputDir: string;
  comment?: string;
  migrated?: boolean;
}

export class AppBuilder {
  config: AppBuilderProps;

  code: CodeMaker;

  constructor(props: AppBuilderProps) {
    this.config = props;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    const { comment, outputFile, imports, appName, stack, outputDir, stages } = this.config;

    this.code.openFile(outputFile);
    if (comment) {
      this.code.line(comment);
      this.code.line();
    }

    imports.render(this.code);

    this.code.line("const app = new App();");

    stages.forEach((stage) => {
      this.code.line(
        `new ${appName.pascal}(app, "${appName.pascal}-${stage}", { stack: "${stack.kebab}", stage: "${stage}" });`
      );
    });

    this.code.closeFile(outputFile);
    await this.code.save(outputDir);
  }
}

export const constructApp = async (props: AppBuilderProps): Promise<void> => {
  const builder = new AppBuilder(props);
  await builder.constructCdkFile();
};
