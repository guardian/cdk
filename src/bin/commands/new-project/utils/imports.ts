import type { CodeMaker } from "codemaker";
import type { Name } from "./utils";

interface Import {
  types: string[];
  components: string[];
  basic?: boolean;
}

export class Imports {
  private readonly imports: Record<string, Import>;

  private constructor(imports: Record<string, Import>) {
    this.imports = imports;
  }

  public static newStackImports(addCfnIncludeImports: boolean): Imports {
    return new Imports({
      ...(addCfnIncludeImports && {
        path: {
          types: [],
          components: ["join"],
        },
        "aws-cdk-lib/cloudformation-include": {
          types: [],
          components: ["CfnInclude"],
        },
      }),
      "aws-cdk-lib": {
        types: ["App"],
        components: [],
      },
      "@guardian/cdk/lib/constructs/core": {
        types: ["GuStackProps"],
        components: ["GuStack"],
      },
    });
  }

  public static newAppImports({ kebab, pascal }: Name): Imports {
    return new Imports({
      "aws-cdk-lib": {
        types: [],
        components: ["App"],
      },
      "source-map-support/register": {
        basic: true,
        types: [],
        components: [],
      },
      [`../lib/${kebab}`]: {
        types: [],
        components: [pascal],
      },
    });
  }

  public static newTestImports({ kebab, pascal }: Name): Imports {
    return new Imports({
      "aws-cdk-lib": {
        types: [],
        components: ["App"],
      },
      "aws-cdk-lib/assertions": {
        types: [],
        components: ["Template"],
      },
      [`./${kebab}`]: {
        types: [],
        components: [pascal],
      },
    });
  }

  render(code: CodeMaker): void {
    Object.entries(this.imports)
      // Render "basic" imports before any others, still in alphabetical order
      // Render relative imports after absolute imports
      .sort(([aKey, aImports], [bKey, bImports]) => {
        if (aImports.basic && !bImports.basic) {
          return -1;
        } else if (bImports.basic && !aImports.basic) {
          return 1;
        } else if (aKey.startsWith(".") && !bKey.startsWith(".")) {
          return 1;
        } else if (bKey.startsWith(".") && !aKey.startsWith(".")) {
          return -1;
        } else {
          return aKey.localeCompare(bKey);
        }
      })
      .forEach(([lib, imports]) => {
        imports.basic && code.line(`import "${lib}";`);

        imports.types.length && code.line(`import type { ${imports.types.sort().join(", ")} } from "${lib}";`);

        imports.components.length && code.line(`import { ${imports.components.sort().join(", ")} } from "${lib}";`);
      });
    code.line();
  }
}
