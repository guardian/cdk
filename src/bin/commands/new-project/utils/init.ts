import { copyFileSync, existsSync, lstatSync, mkdirSync, readdirSync, writeFileSync } from "fs";
import { join } from "path";
import { LibraryInfo } from "../../../../constants";

export interface InitConfig {
  outputDir: string;
}

// TODO: Add yarn or npm flag
// TODO: Add project name flag
export class ProjectBuilder {
  templateDir = `${__dirname}/../template`;
  config: InitConfig;

  static validateConfig = (config: InitConfig): void => {
    const path = config.outputDir;
    if (existsSync(path) && readdirSync(path).length > 0) {
      throw new Error(`Directory ${path} is not empty`);
    }
  };

  constructor(config: InitConfig) {
    this.config = config;
  }

  buildDirectory(): void {
    ProjectBuilder.validateConfig(this.config);

    if (!existsSync(this.config.outputDir)) {
      console.log(`Creating ${this.config.outputDir}`);
      mkdirSync(this.config.outputDir);
    }

    console.log("Creating package.json");
    createPackageJson(this.config.outputDir);

    console.log("Copying template files");
    // TODO: Replace any params in files with .template extensions
    this.copyFiles(this.templateDir, this.config.outputDir);

    console.log("Success!");
  }

  copyFiles(sourcePath: string, targetPath: string): void {
    for (const file of readdirSync(sourcePath)) {
      const path = join(sourcePath, file);

      if (path.endsWith(".ignore")) {
        continue;
      } else if (lstatSync(path).isDirectory()) {
        const nestedTargetPath = join(targetPath, file);
        if (!existsSync(nestedTargetPath)) {
          mkdirSync(nestedTargetPath);
        }
        this.copyFiles(path, nestedTargetPath);
      } else if (path.endsWith(".template")) {
        copyFileSync(path, join(targetPath, file.replace(".template", "")));
      } else {
        copyFileSync(path, join(targetPath, file));
      }
    }
  }
}

export const buildDirectory = (config: InitConfig): void => {
  const builder = new ProjectBuilder(config);
  builder.buildDirectory();
};

function createPackageJson(outputDirectory: string): void {
  const { NODE_ENV, CI } = process.env;
  const isTest = NODE_ENV?.toUpperCase() === "TEST" || CI?.toUpperCase() === "TRUE";

  const contents = {
    name: "cdk",
    version: "0.0.0",
    private: true,
    bin: {
      cdk: "bin/cdk.js",
    },
    scripts: {
      build: "tsc",
      watch: "tsc -w",
      test: "jest",
      "test:dev": "jest --watch",
      format: 'prettier --write "{lib,bin}/**/*.ts"',
      cdk: "cdk",
      lint: "eslint lib/** bin/** --ext .ts --no-error-on-unmatched-pattern",
      synth: "cdk synth --path-metadata false --version-reporting false",
      diff: "cdk diff --path-metadata false --version-reporting false",
    },
    devDependencies: {
      "@aws-cdk/assert": LibraryInfo.AWS_CDK_VERSION,
      "@guardian/eslint-config-typescript": "^0.7.0",
      "@types/jest": "^27.0.2",
      "@types/node": "16.11.7",
      "aws-cdk": LibraryInfo.AWS_CDK_VERSION,
      eslint: "^7.32.0",
      jest: "^27.3.1",
      prettier: "^2.4.1",
      "ts-jest": "^27.0.7",
      "ts-node": "^10.4.0",
      typescript: "~4.4.3",
    },
    dependencies: {
      "@aws-cdk/cloudformation-include": LibraryInfo.AWS_CDK_VERSION,
      "@aws-cdk/core": LibraryInfo.AWS_CDK_VERSION,

      /*
      Do not add `@guardian/cdk` to the generated `package.json` file when in TEST as we'll `npm link` it instead.
      See https://docs.npmjs.com/cli/v8/commands/npm-link#caveat

      TODO remove this once the `new` command allows opting out of automatic dependency installation
       */
      ...(!isTest && { "@guardian/cdk": LibraryInfo.VERSION }),

      "source-map-support": "^0.5.20",
    },
  };
  writeFileSync(`${outputDirectory}/package.json`, JSON.stringify(contents, null, 2));
}
