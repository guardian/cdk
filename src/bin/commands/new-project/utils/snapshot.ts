import type { PackageManager } from "aws-sdk/clients/ecr";
import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

export interface TestBuilderProps {
  imports: Imports;
  appName: Name;
  stackName: Name;
  outputFile: string;
  outputDir: string;
  comment?: string;
  packageManager: PackageManager;
}

export class TestBuilder {
  config: TestBuilderProps;
  imports: Imports;

  code: CodeMaker;

  constructor(props: TestBuilderProps) {
    this.config = props;
    this.imports = props.imports;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    this.code.openFile(this.config.outputFile);
    if (this.config.comment) {
      this.code.line(this.config.comment);
      this.code.line();
    }

    this.config.imports.render(this.code);

    this.config.packageManager === "deno" ? this.addTestDeno() : this.addTest();

    this.code.closeFile(this.config.outputFile);
    await this.code.save(this.config.outputDir);
  }

  addTest(): void {
    const { appName, stackName } = this.config;

    this.code.openBlock(`describe("The ${appName.pascal} stack", () =>`);
    this.code.openBlock(`it("matches the snapshot", () =>`);

    this.code.line("const app = new App();");
    this.code.line(
      `const stack = new ${appName.pascal}(app, "${appName.pascal}", { stack: "${stackName.kebab}", stage: "TEST" });`,
    );

    this.code.line(`const template = Template.fromStack(stack);`);
    this.code.line(`expect(template.toJSON()).toMatchSnapshot();`);

    this.code.closeBlock("});");
    this.code.closeBlock("});");
  }

  addTestDeno(): void {
    const { appName, stackName } = this.config;

    this.code.openBlock(`Deno.test("The ${appName.pascal} stack matches the screenshot", async (ctx) =>`);

    this.code.line("const app = new App();");
    this.code.line(
      `const stack = new ${appName.pascal}(app, "${appName.pascal}", { stack: "${stackName.kebab}", stage: "TEST" });`,
    );

    this.code.line(`const template = Template.fromStack(stack);`);
    this.code.line(`await assertSnapshot(ctx, template.toJSON());`);

    this.code.closeBlock("});");
  }
}

export const constructTest = async (props: TestBuilderProps): Promise<void> => {
  const builder = new TestBuilder(props);
  await builder.constructCdkFile();
};
