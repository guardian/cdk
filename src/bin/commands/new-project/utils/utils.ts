import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
