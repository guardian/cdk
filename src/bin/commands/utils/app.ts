import { CodeMaker } from "codemaker";
import type { Imports } from "./imports";
import type { Name } from "./utils";

interface AppBuilderProps {
  imports: Imports;
  appName: Name;
  stack: Name;
  outputFile: string;
  outputDir: string;
  comment?: string;
  migrated?: boolean;
}

export class AppBuilder {
  config: AppBuilderProps;

  code: CodeMaker;

  constructor(props: AppBuilderProps) {
    this.config = props;

    this.code = new CodeMaker({ indentationLevel: 2 });
    this.code.closeBlockFormatter = (s?: string): string => s ?? "}";
  }

  async constructCdkFile(): Promise<void> {
    this.code.openFile(this.config.outputFile);
    if (this.config.comment) {
      this.code.line(this.config.comment);
      this.code.line();
    }

    this.config.imports.render(this.code);

    this.code.line("const app = new App();");

    this.code.line("const cloudFormationStackName = process.env.GU_CFN_STACK_NAME;");

    this.code.line(
      `new ${this.config.appName.pascal}(app, "${this.config.appName.pascal}", { stack: "${this.config.stack.kebab}", cloudFormationStackName });`
    );

    this.code.closeFile(this.config.outputFile);
    await this.code.save(this.config.outputDir);
  }
}

export const constructApp = async (props: AppBuilderProps): Promise<void> => {
  const builder = new AppBuilder(props);
  await builder.constructCdkFile();
};
