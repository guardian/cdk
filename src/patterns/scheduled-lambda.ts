import { Rule } from "@aws-cdk/aws-events";
import { LambdaFunction } from "@aws-cdk/aws-events-targets";
import { GuLambdaFunction } from "../constructs/lambda";
import type { GuLambdaErrorPercentageMonitoringProps, NoMonitoring } from "../constructs/cloudwatch";
import type { GuStack } from "../constructs/core";
import type { GuFunctionProps } from "../constructs/lambda";
import type { Schedule } from "@aws-cdk/aws-events";

/**
 * Configuration options for the [[`GuScheduledLambda`]] pattern.
 *
 * For all lambda function configuration options, see [[`GuFunctionProps`]].
 *
 * The `schedule` property determines when your lambda is invoked. For example, to invoke
 * the lambda every 5 minutes, use:
 * ```typescript
 * import { Schedule } from "@aws-cdk/aws-events";
 * import { Duration } from "@aws-cdk/core";
 *
 * const props = {
 *   // Other props here
 *   schedule: Schedule.rate(Duration.minutes(5)),
 * }
 * ```
 *
 * To invoke the lambda every weekday at 8am, use:
 * ```
 * import { Schedule } from "@aws-cdk/aws-events";
 *
 * const props = {
 *   // Other props here
 *   schedule: Schedule.expression("cron(0 8 ? * MON-FRI *)"),
 * }
 * ```
 *
 * It is advisable to configure an alarm based on the lambda's error percentage.
 * To do this, add the `monitoringConfiguration` property. The required properties for this are:
 *
 * ```typescript
 * monitoringConfiguration: {
 *   toleratedErrorPercentage: <sensible_error_percentage_threshold>,
 *   snsTopicName: "my-topic-for-cloudwatch-alerts",
 * }
 * ```
 * Other alarm properties (e.g. alarm name and description) will be pre-populated with sensible defaults.
 * For a full list of optional properties, see [[`GuLambdaErrorPercentageMonitoringProps`]].
 *
 * If your team do not use CloudWatch, it's possible to opt-out with the following configuration:
 * ```typescript
 *  monitoringConfiguration: { noMonitoring: true }
 * ```
 */
export interface GuScheduledLambdaProps extends Omit<GuFunctionProps, "errorPercentageMonitoring"> {
  rules: Array<{
    schedule: Schedule;
    description?: string;
  }>;
  monitoringConfiguration: NoMonitoring | GuLambdaErrorPercentageMonitoringProps;
}

/**
 * Pattern which creates all of the resources needed to invoke a lambda function on a schedule.
 *
 * For all configuration options, see [[`GuScheduledLambdaProps`]].
 */
export class GuScheduledLambda extends GuLambdaFunction {
  constructor(scope: GuStack, id: string, props: GuScheduledLambdaProps) {
    const lambdaProps: GuFunctionProps = {
      ...props,
      errorPercentageMonitoring: props.monitoringConfiguration.noMonitoring ? undefined : props.monitoringConfiguration,
    };
    super(scope, id, lambdaProps);

    props.rules.forEach((rule, index) => {
      const target = new LambdaFunction(this);
      new Rule(this, `${id}-${rule.schedule.expressionString}-${index}`, {
        schedule: rule.schedule,
        targets: [target],
        ...(rule.description && { description: rule.description }),
        enabled: true,
      });
    });
  }
}
