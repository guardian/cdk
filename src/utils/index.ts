export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
