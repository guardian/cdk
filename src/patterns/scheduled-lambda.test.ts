import { Duration } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { Schedule } from "aws-cdk-lib/aws-events";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import type { NoMonitoring } from "../constructs/cloudwatch";
import { simpleGuStackForTesting } from "../utils/test";
import { GuScheduledLambda } from "./scheduled-lambda";

describe("The GuScheduledLambda pattern", () => {
  it("should create the correct resources with minimal config", () => {
    const stack = simpleGuStackForTesting();
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      rules: [{ schedule: Schedule.rate(Duration.minutes(1)) }],
      monitoringConfiguration: noMonitoring,
      app: "testing",
    };
    new GuScheduledLambda(stack, "my-lambda-function", props);
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("should create an alarm if monitoring configuration is provided", () => {
    const stack = simpleGuStackForTesting();
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      rules: [{ schedule: Schedule.rate(Duration.minutes(1)) }],
      monitoringConfiguration: {
        toleratedErrorPercentage: 99,
        snsTopicName: "alerts-topic",
      },
      app: "testing",
    };
    new GuScheduledLambda(stack, "my-lambda-function", props);
    Template.fromStack(stack).resourceCountIs("AWS::CloudWatch::Alarm", 1);
  });
});
