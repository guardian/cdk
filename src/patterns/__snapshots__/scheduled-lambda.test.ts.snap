// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuScheduledLambda pattern should create the correct resources with minimal config 1`] = `
Object {
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "mylambdafunction8D341B54": Object {
      "DependsOn": Array [
        "mylambdafunctionServiceRoleDefaultPolicy769897D4",
        "mylambdafunctionServiceRoleE82C2E25",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "lambda.zip",
        },
        "FunctionName": "my-lambda-function",
        "Handler": "my-lambda/handler",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "mylambdafunctionServiceRoleE82C2E25",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "mylambdafunctionServiceRoleDefaultPolicy769897D4": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "mylambdafunctionServiceRoleDefaultPolicy769897D4",
        "Roles": Array [
          Object {
            "Ref": "mylambdafunctionServiceRoleE82C2E25",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "mylambdafunctionServiceRoleE82C2E25": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "mylambdafunctionmylambdafunctionrate1minute06AD0015D": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(1 minute)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "mylambdafunction8D341B54",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "mylambdafunctionmylambdafunctionrate1minute0AllowEventRuleTestmylambdafunction4858F7ED96852F99": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "mylambdafunction8D341B54",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "mylambdafunctionmylambdafunctionrate1minute06AD0015D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
